File: androidApp/src/main/java/com/luisma/conexiones/android/features/game/componets/GameGrid.kt
package com.luisma.conexiones.android.features.game.componets

import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.offset
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.luisma.conexiones.android.core_ui.components.game_cards.GameCardsContracts
import com.luisma.conexiones.android.core_ui.components.game_cards.GuessedCard
import com.luisma.conexiones.android.core_ui.components.game_cards.GuessedCardAnimationType
import com.luisma.conexiones.android.core_ui.components.game_cards.WordCard
import com.luisma.conexiones.android.core_ui.components.game_cards.WordCardAnimationType
import com.luisma.conexiones.android.core_ui.helpers.getColorByDomainColor
import com.luisma.conexiones.android.core_ui.helpers.getColorCoreByDifficulty
import com.luisma.conexiones.models.game.GameGridRowState
import com.luisma.conexiones.models.game.GameTileAnimationType
import com.luisma.conexiones.models.game.GameWordAnimationType

@Composable
fun GameGrid(
    modifier: Modifier = Modifier,
    cardType: GameCardsContracts,
    canSelectCards: Boolean,
    onTapCard: (column: Int, row: Int) -> Unit,
    dismissWordCardAnimation: (column: Int, row: Int) -> Unit,
    dismissTileCardAnimation: (row: Int) -> Unit,
    gridRowsState: Map<Int, GameGridRowState>,
    notSolvedRowsOnLost: Map<Int, Int>
) {

    fun verticalOffset(row: Int): Dp {
        if (row == 0) {
            return 0.dp
        }
        return (cardType.cardHeight + cardType.cardSpacing) * row
    }

    fun horizontalOffset(column: Int): Dp {
        if (column == 0) {
            return 0.dp
        }

        return (cardType.wordCardWidth + cardType.cardSpacing) * column
    }

    fun mapWordCardAnimation(animation: GameWordAnimationType): WordCardAnimationType {
        return when (animation) {
            GameWordAnimationType.NoAnimation -> WordCardAnimationType.NoAnimation
            GameWordAnimationType.Tap -> WordCardAnimationType.Scale
            GameWordAnimationType.Fail -> WordCardAnimationType.TranslateX
        }
    }

    fun mapGuessedCardAnimation(animation: GameTileAnimationType): GuessedCardAnimationType {
        return when (animation) {
            GameTileAnimationType.NoAnimation -> GuessedCardAnimationType.NoAnimation
            GameTileAnimationType.Reveal -> GuessedCardAnimationType.Scale
        }
    }

    Box(
        modifier = modifier,
    ) {
        gridRowsState.forEach { (rowIdx, rowData) ->
            rowData.distribution.forEachIndexed { columnIdx, dist ->
                WordCard(
                    modifier = Modifier.offset(
                        x = horizontalOffset(columnIdx),
                        y = verticalOffset(rowIdx),
                    ),
                    text = rowData.wordsRow[columnIdx],
                    selected = dist.selected,
                    onTap = onTapCard,
                    disabled = rowData.solved || (!dist.selected && !canSelectCards),
                    column = columnIdx,
                    row = rowIdx,
                    sizeType = cardType,
                    dismissAnimation = dismissWordCardAnimation,
                    animationType = mapWordCardAnimation(dist.should.animationType)
                )
            }
            if (rowData.solved) {
                GuessedCard(
                    modifier = Modifier.offset(
                        x = 0.dp,
                        y = verticalOffset(rowIdx)
                    ),
                    description = rowData.solvedTile.description,
                    words = rowData.solvedTile.wordsJoined,
                    color = getColorByDomainColor(getColorCoreByDifficulty(rowData.solvedTile.difficulty)),
                    sizeType = cardType,
                    row = rowIdx,
                    dismissAnimation = dismissTileCardAnimation,
                    animationType = mapGuessedCardAnimation(rowData.tileAnimationType),
                    scaleOrder = if (notSolvedRowsOnLost.isEmpty() || notSolvedRowsOnLost[rowIdx] == null) {
                        0
                    } else {
                        notSolvedRowsOnLost[rowIdx]!!
                    }
                )
            }
        }
    }
}

File: iosApp/iosApp/Features/Game/Components/GameGrid.swift
//
//  GameGrid.swift
//  iosApp
//
//  Created by Luis Manuel García Fernández on 11/10/24.
//  Copyright © 2024 orgName. All rights reserved.
//

import shared
import SwiftUI

struct GameGrid: View {
    @EnvironmentObject var vm: GameViewModel

    let cardSize: GameCardsContractsModel

    var body: some View {
        func verticalOffset(row: Int) -> Double {
            if row == 0 {
                return 0
            }

            return (cardSize.cardHeight + cardSize.cardSpacing) * Double(row)
        }

        func horizontalOffset(column: Int) -> Double {
            if column == 0 {
                return 0
            }

            return (cardSize.wordCardWidth + cardSize.cardSpacing) * Double(column)
        }

        func mapWordCardAnimation(_ animation: GameWordAnimationType) -> WordCarAnimationType {
            switch animation {
            case GameWordAnimationType.noanimation:
                return WordCarAnimationType.noAnimation
            case GameWordAnimationType.tap:
                return WordCarAnimationType.scale
            case GameWordAnimationType.fail:
                return WordCarAnimationType.jump
            default:
                return WordCarAnimationType.noAnimation
            }
        }

        func mapGuessedCardAnimation(_ animation: GameTileAnimationType) -> GuessedCardAnimationType {
            switch animation {
            case GameTileAnimationType.noanimation:
                return GuessedCardAnimationType.noAnimation
            case GameTileAnimationType.reveal:
                return GuessedCardAnimationType.scale
            default:
                return GuessedCardAnimationType.noAnimation
            }
        }

        return ZStack(
            alignment: .topLeading
        ) {
            Color.clear
            ForEach(
                Array(vm.state.gridRowsState.keys),
                id: \.self
            ) { rowIdx in
                let rowData = vm.state.gridRowsState[rowIdx]!
                let distribution = rowData.distribution

                ForEach(
                    0 ..< distribution.count,
                    id: \.self
                ) { columnIdx in
                    let dist = distribution[columnIdx]
                    return WordCard(
                        disabled: !dist.selected && !vm.state.canSelectWords,
                        sizeType: cardSize.contract,
                        text: rowData.wordsRow[columnIdx],
                        selected: dist.selected,
                        onTap: { column, row in
                            vm.sendEvent(event: .selectWord(column: column, row: row))
                        },
                        column: columnIdx,
                        row: rowIdx,
                        animationType: mapWordCardAnimation(dist.should.animationType),
                        dismissAnimation: { column, row in
                            vm.sendEvent(event: .dismissAnimation(column: column, row: row))
                        }
                    )
                    .offset(
                        x: horizontalOffset(column: columnIdx),
                        y: verticalOffset(row: rowIdx)
                    )
                }

                if rowData.solved {
                    GuessedCard(
                        row: rowIdx,
                        description: rowData.solvedTile.description_,
                        words: rowData.solvedTile.wordsJoined,
                        color: getColorByDomainColor(color: getColorByDifficulty(difficulty: Int(rowData.solvedTile.difficulty))),
                        sizeType: cardSize.contract,
                        animationType: mapGuessedCardAnimation(rowData.tileAnimationType),
                        scaleOrder: (vm.state.notSolvedRowsOnLost.isEmpty || vm.state.notSolvedRowsOnLost[rowIdx] == nil ? 0 : vm.state.notSolvedRowsOnLost[rowIdx]) ?? 0,
                        dissmissAnimation: { row in vm.sendEvent(event: .dismissTileAnimation(row: row)) }
                    )
                    .offset(
                        x: 0,
                        y: verticalOffset(row: rowIdx)
                    )
                }
            }
        }
    }
}
